### DISCUSSION TITLE
Mistake about compare two Bluetooth.UUID() instance.

### DISCUSSION THREAD

**Discussion by walklinewang** [micropython/micropython#15136] [ESP32] [answered]
I'm trying to discover characteristic's descriptor, and passing characteristic's start_handle & end_handle to [bluetooth.BLE.gattc_discover_descriptors](https://docs.micropython.org/en/latest/library/bluetooth.html#bluetooth.BLE.gattc_discover_descriptors) function.

I could receive many `uuid` in `GATTC_DESCRIPTOR_RESULT` callback, but there are mostly useless besides the descriptors's UUID, such like `0x2803` and characteristic's UUID.

For this reason, I want to filter them out, i.e. values other than `0x2900` and `0x2911`, but I found this issue:

```python
from bluetooth import UUID

a = UUID(0x2803)
b = UUID(0x2900)
c = UUID(0x2911)

a > b # True   ?????
a > c # False

```

My question is how to fix this issue? Or how to receive discriptor correctly?

My mpy version `MicroPython v1.23.0-preview.360.gc3301da17 on 2024-05-13`, idf version `v5.0.4`

Update:

It looks as if it swaps the high and low bytes when comparing two UUIDs, this code gets the right result.

```python
from bluetooth import UUID

a = UUID(0x0328)
b = UUID(0x0029)
c = UUID(0x1129)

a > b # False
a > c # False

```

**Comment by redhead-p** [comment]
I've written BLE code using these functions without problem.  I use code that tests for equality between the returned UUID and the one's I want to test for.  Using > & < for UUIDs and the like is problematic unless the behaviour of these operators is explicitly defined as it is for Strings say.

Regards

Paul

**Reply by walklinewang** [reply]
> unless the behaviour of these operators is explicitly defined as it is for Strings say.

In fact, the [bluetooth_uuid_binary_op()](https://github.com/micropython/micropython/blob/30a9ccf4caa72c62cb8656a1572518fef34b08a4/extmod/modbluetooth.c#L180) is acting like a string, it used `memcmp()` to compare two UUID's data[], but unfortunately, for the type of `MP_BLUETOOTH_UUID_TYPE_16` UUID, first element in data[] is low byte of the value, and second is high ([here](https://github.com/micropython/micropython/blob/30a9ccf4caa72c62cb8656a1572518fef34b08a4/extmod/modbluetooth.c#L111)).

I bingged the source code of `memcmp`, looks like it will returned when the subtraction of two numbers is not zero.

```c
int memcmp(const void *s1, const void *s2, size_t n)
{
	const unsigned char *su1, *su2;
	int res = 0;

	for (su1 = s1, su2 = s2; 0 < n; su1++, su2++, n--)
		if ((res = *su1 - *su2) != 0)
			break;
	return res;
}
```

I tried to modify the source code:

```c
self->data[1] = value & 0xff;
self->data[0] = (value >> 8) & 0xff;
```

But that's not how the other related code works.

**Reply by redhead-p** [reply]
Hmm OK - but I don't really understand why you want to know why one UUID is bigger than another.  What does it mean in the context of BLE?

**Reply by walklinewang** [reply]
I'm writing a scanner class used to discover BLE device‘s services, characteristics and descriptors, in this while I found an issue:

> I could receive many uuid in GATTC_DESCRIPTOR_RESULT callback, but there are mostly useless besides the descriptors's UUID, such like 0x2803 and characteristic's UUID.

The descriptor's UUID are in range 0x2900~0x2911 (defined by Bluetooth SIG), when I want to discover a characteristic's descriptors, I will receive many out of range UUIDs, I just want to record the correct UUID and filter others. 😅

**Reply by redhead-p** [reply]
OK - I hadn't spotted the bit about the range check but I wouldn't bother with that - just check for the UUIDs you are interested in. E.g.

```
from bluetooth import UUID

a = UUID(0x0328)
b = UUID(0x0029)
c = UUID(0x1129)


def _gattc_descriptor_result(self, data):
    _, descriptor_handle,  uuid = data

    if uuid == a:
        ...
    elif uuid == b:
        ...
    elif uuid == c:
        ...
```

If you don't like long 'if then' constructs (and I don't) then here's an alternative approach taken from the way I deal with descriptors:

```
# Characteristic look up
_chUUIDlu = {ID_UUID: 'i', STATE_UUID: 's', CMD_UUID: 'c', TAGID_UUID: 't', TAG_DATA_UUID: 'd',
             DEV_NAME_UUID: 'N'}

...

    def _gattc_characteristic_result(self, data):
        # a characteristic report - add to characteristic table if UUID of known characteristic
        _, end_handle, value_handle, properties, uuid = data
        try:
            self._characteristics[_chUUIDlu[uuid]] = (end_handle, value_handle, properties)
        except KeyError:
            pass
```

This makes the underlying dictionary code workout what's valid and what isn't.

Paul

**Reply by walklinewang** [reply]
Thanks for your demo, I'll try this later.

**Comment by walklinewang** [comment]
Just for now.

```python
# if UUID(0x2900) <= uuid <= UUID(0x2911):
#    print(uuid)
if uuid == UUID(0x2900) or\
   uuid == UUID(0x2901) or\
   uuid == UUID(0x2902) or\
   uuid == UUID(0x2903) or\
   uuid == UUID(0x2904) or\
   uuid == UUID(0x2905) or\
   uuid == UUID(0x2906) or\
   uuid == UUID(0x2907) or\
   uuid == UUID(0x2908) or\
   uuid == UUID(0x2909) or\
   uuid == UUID(0x290a) or\
   uuid == UUID(0x290b) or\
   uuid == UUID(0x290c) or\
   uuid == UUID(0x290d) or\
   uuid == UUID(0x290e) or\
   uuid == UUID(0x2910) or\
   uuid == UUID(0x2911):
    pass
```

**Reply by walklinewang** [reply, marked as answer]
Update:

```python
DESCRIPTORS_UUID = [
    UUID(0x2900), UUID(0x2901), UUID(0x2902), UUID(0x2903), UUID(0x2904),
    UUID(0x2905), UUID(0x2906), UUID(0x2907), UUID(0x2908), UUID(0x2909),
    UUID(0x290a), UUID(0x290b), UUID(0x290c), UUID(0x290d), UUID(0x290e),
    UUID(0x2910), UUID(0x2911)
]

if uuid in DESCRIPTORS_UUID:
    pass
```
